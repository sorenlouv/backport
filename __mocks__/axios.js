const axios = jest.genMockFromModule('axios');
axios.get = jest.fn(() => {
  throw new Error('Forgot to mock axios.get');
});
axios.post = jest.fn(() => {
  throw new Error('Forgot to mock axios.post');
});
module.exports = axios;
