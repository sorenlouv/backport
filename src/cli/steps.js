const github = require('../lib/github');
const {
  getCommitByPrompt,
  getCommitBySha,
  getBranchesByPrompt,
  doBackportVersions,
  handleErrors,
  maybeSetupRepo
} = require('./cliService');

function initSteps(options) {
  const [owner, repoName] = options.upstream.split('/');

  let commits, branches;
  github.setAccessToken(options.accessToken);

  const promise = options.sha
    ? getCommitBySha({ owner, repoName, sha: options.sha })
    : getCommitByPrompt({
        owner,
        repoName,
        author: options.own ? options.username : null,
        multipleCommits: options.multipleCommits
      });

  return promise
    .then(c => {
      commits = c;
    })
    .then(() => getBranchesByPrompt(options.branches, options.multipleBranches))
    .then(v => {
      branches = v;
    })
    .then(() => maybeSetupRepo(owner, repoName, options.username))
    .then(() =>
      doBackportVersions({
        owner,
        repoName,
        commits,
        branches,
        username: options.username,
        labels: options.labels
      })
    )
    .catch(handleErrors);
}

module.exports = initSteps;
