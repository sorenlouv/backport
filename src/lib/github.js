const axios = require('axios');
const querystring = require('querystring');
const get = require('lodash.get');
const constants = require('./constants');

let accessToken;
function getCommitMessage(message) {
  return message.split('\n')[0].trim();
}

function getCommits(owner, repoName, author) {
  const urlArgs = {
    per_page: 20,
    access_token: accessToken
  };

  if (author) {
    urlArgs.author = author;
    urlArgs.per_page = 5;
  }

  return axios(
    `https://api.github.com/repos/${owner}/${repoName}/commits?${querystring.stringify(
      urlArgs
    )}`
  )
    .catch(handleError)
    .then(res =>
      res.data.map(commit => {
        const message = getCommitMessage(commit.commit.message);
        return {
          message,
          sha: commit.sha
        };
      })
    );
}

function getCommit(owner, repoName, sha) {
  return axios(
    `https://api.github.com/repos/${owner}/${repoName}/commits/${sha}?access_token=${accessToken}`
  )
    .catch(handleError)
    .then(res => ({
      message: res.data.commit.message,
      sha: res.data.sha
    }));
}

function createPullRequest(owner, repoName, payload) {
  return axios
    .post(
      `https://api.github.com/repos/${owner}/${repoName}/pulls?access_token=${accessToken}`,
      payload
    )
    .catch(handleError);
}

function addLabels(owner, repoName, pullNumber, labels) {
  return axios
    .post(
      `https://api.github.com/repos/${owner}/${repoName}/issues/${pullNumber}/labels?access_token=${accessToken}`,
      labels
    )
    .catch(handleError);
}

function getPullRequestByCommit(owner, repoName, commitSha) {
  return axios(
    `https://api.github.com/search/issues?q=repo:${owner}/${repoName}+${commitSha}&access_token=${accessToken}`
  )
    .catch(handleError)
    .then(res => get(res.data.items[0], 'number'));
}

function setAccessToken(_accessToken) {
  accessToken = _accessToken;
}

class GithubError extends Error {
  constructor(message) {
    super();
    Error.captureStackTrace(this, GithubError);
    this.code = constants.GITHUB_ERROR;
    this.message = message;
  }
}

function handleError(e) {
  if (get(e.response, 'data')) {
    throw new GithubError(e.response.data);
  }

  throw e;
}

module.exports = {
  setAccessToken,
  addLabels,
  createPullRequest,
  getCommit,
  getCommits,
  getPullRequestByCommit
};
