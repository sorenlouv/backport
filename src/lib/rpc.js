const promisify = require('es6-promisify');
const process = require('child_process');
const fs = require('fs');
const mkdirp = require('mkdirp');
const findUp = require('find-up');

module.exports = {
  exec: promisify(process.exec),
  writeFile: promisify(fs.writeFile),
  readFile: promisify(fs.readFile),
  stat: promisify(fs.stat),
  statSync: fs.statSync,
  mkdirp: promisify(mkdirp),
  findUp
};
