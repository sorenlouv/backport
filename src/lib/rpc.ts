import { promisify } from 'util';
import { exec as execLegacy } from 'child_process';
import fs from 'fs';
import mkdirpLegacy from 'mkdirp';

const execAsPromised = promisify(execLegacy);
const mkdirpAsPromised = promisify(mkdirpLegacy);

export function exec(cmd: string, options?: {}) {
  return execAsPromised(cmd, { maxBuffer: 100 * 1024 * 1024, ...options });
}
export const writeFile = promisify(fs.writeFile);
export const readFile = promisify(fs.readFile);
export const stat = promisify(fs.stat);
export const chmod = promisify(fs.chmod);
export const statSync = fs.statSync;
export const mkdirp = (path: string) => mkdirpAsPromised(path);
