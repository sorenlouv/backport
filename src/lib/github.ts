import {
  GithubQuery,
  GithubCommit,
  PullRequest,
  Commit,
  GithubIssue,
  GithubPullRequestPayload,
  GithubIssues,
  GithubApiError
} from '../types/types';

import axios, { AxiosResponse } from 'axios';
import querystring from 'querystring';
import get from 'lodash.get';
import { HandledError } from './errors';

let accessToken: string;
function getCommitMessage(message: string) {
  return message.split('\n')[0].trim();
}

export async function getCommits(
  owner: string,
  repoName: string,
  author: string | null
): Promise<Commit[]> {
  const query: GithubQuery = {
    access_token: accessToken,
    per_page: 20
  };

  if (author) {
    query.author = author;
    query.per_page = 5;
  }

  try {
    const res: AxiosResponse<GithubCommit[]> = await axios(
      `https://api.github.com/repos/${owner}/${repoName}/commits?${querystring.stringify(
        query
      )}`
    );

    const promises = res.data.map(async commit => {
      const sha = commit.sha;
      return {
        message: getCommitMessage(commit.commit.message),
        sha,
        pullRequest: await getPullRequestBySha(owner, repoName, sha)
      };
    });

    return Promise.all(promises);
  } catch (e) {
    throw getError(e);
  }
}

export async function getCommit(
  owner: string,
  repoName: string,
  sha: string
): Promise<Commit> {
  try {
    const res: AxiosResponse<GithubCommit> = await axios(
      `https://api.github.com/repos/${owner}/${repoName}/commits/${sha}?access_token=${accessToken}`
    );
    const fullSha = res.data.sha;
    const pullRequest = await getPullRequestBySha(owner, repoName, fullSha);

    return {
      message: getCommitMessage(res.data.commit.message),
      sha: fullSha,
      pullRequest
    };
  } catch (e) {
    throw getError(e);
  }
}

export async function createPullRequest(
  owner: string,
  repoName: string,
  payload: GithubPullRequestPayload
): Promise<PullRequest> {
  try {
    const res: AxiosResponse<GithubIssue> = await axios.post(
      `https://api.github.com/repos/${owner}/${repoName}/pulls?access_token=${accessToken}`,
      payload
    );
    return {
      html_url: res.data.html_url,
      number: res.data.number
    };
  } catch (e) {
    throw getError(e);
  }
}

export async function addLabels(
  owner: string,
  repoName: string,
  pullNumber: number,
  labels: string[]
) {
  try {
    return await axios.post(
      `https://api.github.com/repos/${owner}/${repoName}/issues/${pullNumber}/labels?access_token=${accessToken}`,
      labels
    );
  } catch (e) {
    throw getError(e);
  }
}

async function getPullRequestBySha(
  owner: string,
  repoName: string,
  commitSha: string
): Promise<number> {
  try {
    const res: AxiosResponse<GithubIssues> = await axios(
      `https://api.github.com/search/issues?q=repo:${owner}/${repoName}+${commitSha}+base:master&access_token=${accessToken}`
    );
    return get(res.data.items[0], 'number');
  } catch (e) {
    throw getError(e);
  }
}

export function setAccessToken(token: string) {
  accessToken = token;
}

function getError(e: GithubApiError) {
  if (e.response && e.response.data) {
    return new HandledError(
      JSON.stringify({ ...e.response.data, axiosUrl: e.config.url }, null, 4)
    );
  }

  return e;
}
