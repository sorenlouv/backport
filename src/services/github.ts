import {
  GithubApiError,
  GithubCommit,
  GithubIssue,
  GithubQuery,
  GithubSearch
} from '../types/GithubApi';
import axios, { AxiosResponse } from 'axios';
import querystring from 'querystring';
import get from 'lodash.get';
import isEmpty from 'lodash.isempty';
import { HandledError } from './HandledError';
import { getPullRequestPayload } from '../steps/doBackportVersions';

export interface Commit {
  sha: string;
  message: string;
  pullNumber?: number;
}

let accessToken: string;
function getCommitMessage(message: string) {
  return message.split('\n')[0].trim();
}

export async function fetchCommitsByAuthor(
  owner: string,
  repoName: string,
  author: string | null
): Promise<Commit[]> {
  const query: GithubQuery = {
    access_token: accessToken,
    per_page: 10
  };

  if (author) {
    query.author = author;
    query.per_page = 5;
  }

  try {
    const res: AxiosResponse<GithubCommit[]> = await axios(
      `https://api.github.com/repos/${owner}/${repoName}/commits?${querystring.stringify(
        query
      )}`
    );

    const promises = res.data.map(async commit => {
      const sha = commit.sha;
      return {
        message: getCommitMessage(commit.commit.message),
        sha,
        pullNumber: await fetchPullRequestNumberBySha(owner, repoName, sha)
      };
    });

    return Promise.all(promises);
  } catch (e) {
    throw getError(e);
  }
}

export async function fetchCommitBySha(
  owner: string,
  repoName: string,
  sha: string
): Promise<Commit> {
  try {
    const res: AxiosResponse<GithubSearch<GithubCommit>> = await axios(
      `https://api.github.com/search/commits?q=hash:${sha}%20repo:${owner}/${repoName}&per_page=1&access_token=${accessToken}`,
      {
        headers: {
          Accept: 'application/vnd.github.cloak-preview'
        }
      }
    );

    if (isEmpty(res.data.items)) {
      throw new HandledError(`No commit found for SHA: ${sha}`);
    }

    const commitRes = res.data.items[0];
    const fullSha = commitRes.sha;
    const pullNumber = await fetchPullRequestNumberBySha(
      owner,
      repoName,
      fullSha
    );

    return {
      message: getCommitMessage(commitRes.commit.message),
      sha: fullSha,
      pullNumber
    };
  } catch (e) {
    throw getError(e);
  }
}

async function fetchPullRequestNumberBySha(
  owner: string,
  repoName: string,
  commitSha: string
): Promise<number> {
  try {
    const res: AxiosResponse<GithubSearch<GithubIssue>> = await axios(
      `https://api.github.com/search/issues?q=repo:${owner}/${repoName}+${commitSha}+base:master&access_token=${accessToken}`
    );
    return get(res.data.items[0], 'number');
  } catch (e) {
    throw getError(e);
  }
}

export async function createPullRequest(
  owner: string,
  repoName: string,
  payload: ReturnType<typeof getPullRequestPayload>
) {
  try {
    const res: AxiosResponse<GithubIssue> = await axios.post(
      `https://api.github.com/repos/${owner}/${repoName}/pulls?access_token=${accessToken}`,
      payload
    );
    return {
      html_url: res.data.html_url,
      number: res.data.number
    };
  } catch (e) {
    throw getError(e);
  }
}

export async function addLabelsToPullRequest(
  owner: string,
  repoName: string,
  pullNumber: number,
  labels: string[]
) {
  try {
    return await axios.post(
      `https://api.github.com/repos/${owner}/${repoName}/issues/${pullNumber}/labels?access_token=${accessToken}`,
      labels
    );
  } catch (e) {
    throw getError(e);
  }
}

export function setAccessToken(token: string) {
  accessToken = token;
}

function getError(e: GithubApiError) {
  if (e.response && e.response.data) {
    return new HandledError(
      JSON.stringify({ ...e.response.data, axiosUrl: e.config.url }, null, 4)
    );
  }

  return e;
}
