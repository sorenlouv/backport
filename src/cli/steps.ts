import isEmpty from 'lodash.isempty';
import { setAccessToken } from '../lib/github';

import {
  getCommitsByPrompt,
  getCommitBySha,
  getBranchesByPrompt,
  doBackportVersions,
  handleErrors,
  maybeSetupRepo
} from './cliService';
import { BackportOptions, BranchChoice } from '../types/types';

export async function initSteps(options: BackportOptions) {
  const [owner, repoName] = options.upstream.split('/');
  setAccessToken(options.accessToken);

  try {
    const author = options.all ? null : options.username;
    const commits = options.sha
      ? [await getCommitBySha(owner, repoName, options.sha)]
      : await getCommitsByPrompt(
          owner,
          repoName,
          author,
          options.multipleCommits
        );

    const branches = !isEmpty(options.branches)
      ? (options.branches as string[])
      : await getBranchesByPrompt(
          options.branchChoices as BranchChoice[],
          options.multipleBranches
        );

    await maybeSetupRepo(owner, repoName, options.username);
    await doBackportVersions(
      owner,
      repoName,
      commits,
      branches,
      options.username,
      options.labels
    );
  } catch (e) {
    handleErrors(e);
  }
}
