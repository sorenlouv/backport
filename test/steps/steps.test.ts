import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import inquirer from 'inquirer';
import * as childProcess from 'child_process';
import { commitsMock } from '../mocks/commits';
import { initSteps } from '../../src/steps/steps';
import * as github from '../../src/services/github';
import * as rpc from '../../src/services/rpc';

function mockGetPullRequest(
  axiosMock: MockAdapter,
  owner: string,
  repoName: string,
  commitSha: string,
  accessToken: string
) {
  return axiosMock
    .onGet(
      `https://api.github.com/search/issues?q=repo:${owner}/${repoName}+${commitSha}+base:master&access_token=${accessToken}`
    )
    .reply(200, {
      items: [
        {
          number: `Pull number for ${commitSha}`
        }
      ]
    });
}

function mockVerifyAccessToken(
  axiosMock: MockAdapter,
  owner: string,
  repoName: string,
  accessToken: string
) {
  return axiosMock
    .onHead(
      `https://api.github.com/repos/${owner}/${repoName}?access_token=${accessToken}`
    )
    .reply(200);
}

function mockGetCommits(
  axiosMock: MockAdapter,
  {
    owner = 'elastic',
    repoName = 'kibana',
    accessToken = 'myAccessToken',
    author = 'sqren',
    perPage = 5,
    res
  }: {
    owner?: string;
    repoName?: string;
    accessToken?: string;
    author?: string;
    perPage?: number;
    res: {}[];
  }
) {
  return axiosMock
    .onGet(
      `https://api.github.com/repos/${owner}/${repoName}/commits?access_token=${accessToken}&per_page=${perPage}&author=${author}`
    )
    .reply(200, res);
}

function mockCreatePullRequest(
  axiosMock: MockAdapter,
  {
    owner = 'elastic',
    repoName = 'kibana',
    accessToken = 'myAccessToken',
    res
  }: {
    owner?: string;
    repoName?: string;
    accessToken?: string;
    res: any;
  }
) {
  return axiosMock
    .onPost(
      `https://api.github.com/repos/${owner}/${repoName}/pulls?access_token=${accessToken}`
    )
    .reply(200, res);
}

describe('run through steps', () => {
  let axiosMock: MockAdapter;
  let execMock: jest.SpyInstance;
  let inquirerPromptMock: jest.Mock;

  afterEach(() => {
    inquirerPromptMock.mockClear();
    execMock.mockClear();
  });

  beforeEach(async () => {
    const owner = 'elastic';
    const repoName = 'kibana';
    const upstream = `${owner}/${repoName}`;
    const accessToken = 'myAccessToken';

    execMock = jest.spyOn(childProcess, 'exec');

    jest.spyOn(rpc, 'writeFile').mockResolvedValue(undefined);
    jest.spyOn(rpc, 'mkdirp').mockResolvedValue(undefined);

    jest.spyOn(github, 'fetchCommitsByAuthor');
    jest.spyOn(github, 'createPullRequest');

    inquirerPromptMock = jest
      .spyOn(inquirer, 'prompt')
      .mockResolvedValueOnce({
        promptResult: {
          message: 'myCommitMessage',
          sha: 'myCommitSha',
          pullNumber: 'myPullRequestNumber'
        }
      })
      .mockResolvedValueOnce({
        promptResult: '6.2'
      });

    axiosMock = new MockAdapter(axios);

    mockVerifyAccessToken(axiosMock, owner, repoName, accessToken);

    mockGetCommits(axiosMock, {
      owner,
      repoName,
      accessToken,
      res: commitsMock
    });

    mockGetPullRequest(
      axiosMock,
      owner,
      repoName,
      'commitSha',
      'myAccessToken'
    );
    mockGetPullRequest(
      axiosMock,
      owner,
      repoName,
      'commitSha2',
      'myAccessToken'
    );
    mockGetPullRequest(
      axiosMock,
      owner,
      repoName,
      'commitSha3',
      'myAccessToken'
    );

    mockCreatePullRequest(axiosMock, {
      res: {
        html_url: 'myHtmlUrl'
      }
    });

    await initSteps({
      accessToken: 'myAccessToken',
      all: false,
      branches: [],
      branchChoices: [
        { name: '6.x' },
        { name: '6.0' },
        { name: '5.6' },
        { name: '5.5' },
        { name: '5.4' }
      ],
      labels: [],
      multiple: false,
      multipleBranches: false,
      multipleCommits: false,
      prDescription: 'myPrDescription',
      sha: undefined,
      upstream,
      username: 'sqren'
    });
  });

  it('should make correct requests', () => {
    expect(axiosMock.history).toMatchSnapshot();
  });

  it('getCommit should be called with correct args', () => {
    expect(github.fetchCommitsByAuthor).toHaveBeenCalledWith(
      'elastic',
      'kibana',
      'sqren'
    );
  });

  it('createPullRequest should be called with correct args', () => {
    expect(github.createPullRequest).toHaveBeenCalledWith('elastic', 'kibana', {
      base: '6.2',
      body: `Backports the following commits to 6.2:\n - myCommitMessage (#myPullRequestNumber)\n\nmyPrDescription`,
      head: 'sqren:backport/6.2/pr-myPullRequestNumber',
      title: '[6.2] myCommitMessage'
    });
  });

  it('prompt calls should match snapshot', () => {
    expect(inquirer.prompt).toHaveBeenCalledTimes(2);
    expect(inquirerPromptMock.mock.calls).toMatchSnapshot();
  });

  it('exec should be called with correct args', () => {
    expect(execMock).toHaveBeenCalledTimes(9);
    expect(execMock.mock.calls).toMatchSnapshot();
  });
});
