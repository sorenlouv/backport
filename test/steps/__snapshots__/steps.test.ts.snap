// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`run through steps exec should be called with correct args 1`] = `
Array [
  Array [
    "git clone https://myAccessToken@github.com/elastic/kibana.git --progress",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic",
      "maxBuffer": 104857600,
    },
    [Function],
  ],
  Array [
    "git remote rm sqren",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
      "maxBuffer": 104857600,
    },
    [Function],
  ],
  Array [
    "git remote add sqren https://myAccessToken@github.com/sqren/kibana.git",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
      "maxBuffer": 104857600,
    },
    [Function],
  ],
  Array [
    "git remote rm elastic",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
      "maxBuffer": 104857600,
    },
    [Function],
  ],
  Array [
    "git remote add elastic https://myAccessToken@github.com/elastic/kibana.git",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
      "maxBuffer": 104857600,
    },
    [Function],
  ],
  Array [
    "git reset --hard && git clean -d --force && git checkout master && git pull origin master",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
      "maxBuffer": 104857600,
    },
    [Function],
  ],
  Array [
    "git fetch origin 6.2 && git branch backport/6.2/pr-myPullRequestNumber origin/6.2 --force && git checkout backport/6.2/pr-myPullRequestNumber ",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
      "maxBuffer": 104857600,
    },
    [Function],
  ],
  Array [
    "git cherry-pick commitSha",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
      "maxBuffer": 104857600,
    },
    [Function],
  ],
  Array [
    "git push sqren backport/6.2/pr-myPullRequestNumber:backport/6.2/pr-myPullRequestNumber --force",
    Object {
      "cwd": "/myHomeDir/.backport/repositories/elastic/kibana",
      "maxBuffer": 104857600,
    },
    [Function],
  ],
]
`;

exports[`run through steps prompt calls should match snapshot 1`] = `
Array [
  Array [
    Array [
      Object {
        "choices": Array [
          Object {
            "name": "1. Update readme: Make ui-framework url a link",
            "short": "Update readme: Make ui-framework url a link",
            "value": Object {
              "message": "Update readme: Make ui-framework url a link",
              "pullNumber": "Pull number for commitSha",
              "sha": "commitSha",
            },
          },
          Object {
            "name": "2. Update react 15.6.1 (#13672)",
            "short": "Update react 15.6.1 (#13672)",
            "value": Object {
              "message": "Update react 15.6.1 (#13672)",
              "pullNumber": "Pull number for commitSha2",
              "sha": "commitSha2",
            },
          },
          Object {
            "name": "3. Mock out static files when running in Jest (#13315)",
            "short": "Mock out static files when running in Jest (#13315)",
            "value": Object {
              "message": "Mock out static files when running in Jest (#13315)",
              "pullNumber": "Pull number for commitSha3",
              "sha": "commitSha3",
            },
          },
        ],
        "message": "Select commit to backport",
        "name": "promptResult",
        "pageSize": 3,
        "type": "list",
      },
    ],
  ],
  Array [
    Array [
      Object {
        "choices": Array [
          Object {
            "name": "6.x",
          },
          Object {
            "name": "6.0",
          },
          Object {
            "name": "5.6",
          },
          Object {
            "name": "5.5",
          },
          Object {
            "name": "5.4",
          },
        ],
        "message": "Select branch to backport to",
        "name": "promptResult",
        "type": "list",
      },
    ],
  ],
]
`;

exports[`run through steps should make correct requests 1`] = `
Object {
  "delete": Array [],
  "get": Array [
    Object {
      "data": undefined,
      "headers": Object {
        "Accept": "application/json, text/plain, */*",
      },
      "maxContentLength": -1,
      "method": "get",
      "timeout": 0,
      "transformRequest": Object {
        "0": [Function],
      },
      "transformResponse": Object {
        "0": [Function],
      },
      "url": "https://api.github.com/repos/elastic/kibana/commits?access_token=myAccessToken&per_page=5&author=sqren",
      "validateStatus": [Function],
      "xsrfCookieName": "XSRF-TOKEN",
      "xsrfHeaderName": "X-XSRF-TOKEN",
    },
    Object {
      "data": undefined,
      "headers": Object {
        "Accept": "application/json, text/plain, */*",
      },
      "maxContentLength": -1,
      "method": "get",
      "timeout": 0,
      "transformRequest": Object {
        "0": [Function],
      },
      "transformResponse": Object {
        "0": [Function],
      },
      "url": "https://api.github.com/search/issues?q=repo:elastic/kibana+commitSha+base:master&access_token=myAccessToken",
      "validateStatus": [Function],
      "xsrfCookieName": "XSRF-TOKEN",
      "xsrfHeaderName": "X-XSRF-TOKEN",
    },
    Object {
      "data": undefined,
      "headers": Object {
        "Accept": "application/json, text/plain, */*",
      },
      "maxContentLength": -1,
      "method": "get",
      "timeout": 0,
      "transformRequest": Object {
        "0": [Function],
      },
      "transformResponse": Object {
        "0": [Function],
      },
      "url": "https://api.github.com/search/issues?q=repo:elastic/kibana+commitSha2+base:master&access_token=myAccessToken",
      "validateStatus": [Function],
      "xsrfCookieName": "XSRF-TOKEN",
      "xsrfHeaderName": "X-XSRF-TOKEN",
    },
    Object {
      "data": undefined,
      "headers": Object {
        "Accept": "application/json, text/plain, */*",
      },
      "maxContentLength": -1,
      "method": "get",
      "timeout": 0,
      "transformRequest": Object {
        "0": [Function],
      },
      "transformResponse": Object {
        "0": [Function],
      },
      "url": "https://api.github.com/search/issues?q=repo:elastic/kibana+commitSha3+base:master&access_token=myAccessToken",
      "validateStatus": [Function],
      "xsrfCookieName": "XSRF-TOKEN",
      "xsrfHeaderName": "X-XSRF-TOKEN",
    },
  ],
  "head": Array [],
  "list": Array [],
  "options": Array [],
  "patch": Array [],
  "post": Array [
    Object {
      "data": "{\\"title\\":\\"[6.2] myCommitMessage\\",\\"body\\":\\"Backports the following commits to 6.2:\\\\n - myCommitMessage (#myPullRequestNumber)\\",\\"head\\":\\"sqren:backport/6.2/pr-myPullRequestNumber\\",\\"base\\":\\"6.2\\"}",
      "headers": Object {
        "Accept": "application/json, text/plain, */*",
        "Content-Type": "application/json;charset=utf-8",
      },
      "maxContentLength": -1,
      "method": "post",
      "timeout": 0,
      "transformRequest": Object {
        "0": [Function],
      },
      "transformResponse": Object {
        "0": [Function],
      },
      "url": "https://api.github.com/repos/elastic/kibana/pulls?access_token=myAccessToken",
      "validateStatus": [Function],
      "xsrfCookieName": "XSRF-TOKEN",
      "xsrfHeaderName": "X-XSRF-TOKEN",
    },
  ],
  "put": Array [],
}
`;
